mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendEmailAction.php

```diff
@@ -171,40 +186,115 @@ public function run(StepRunArgs $args, StepRunController $controller): void {
+  }
+
+  private function scheduleEmail(StepRunArgs $args, NewsletterEntity $newsletter, SubscriberEntity $subscriber): void {
+    $meta = $this->getNewsletterMeta($args);
+    try {
+      $this->automationEmailScheduler->createSendingTask($newsletter, $subscriber, $meta);
+    } catch (Throwable $e) {
+      throw InvalidStateException::create()->withMessage(__('Could not create sending task.', 'mailpoet'));
+    }
+  }
+
+  private function updateRunLogData(StepRunArgs $args, string $key, mixed $value): void {
+    $run = $args->getAutomationRun();
+    $step = $args->getStep();
+    $storage = new AutomationRunLogStorage();
+
+    $runLog = $storage->getAutomationRunLogByRunAndStepId($run->getId(), $step->getId());
+    if ($runLog === null) {
+      $runLog = new AutomationRunLog($run->getId(), $step->getId(), $step->getType());
+    }
+    $runLog->setData($key, $value);
+
+    if ($runLog->getId() !== null) {
+      $storage->updateAutomationRunLog($runLog);
+    } else {
+      $storage->createAutomationRunLog($runLog);
+    }
+  }
+
+  private function getRunLogData(StepRunArgs $args): array {
+    $run = $args->getAutomationRun();
+    $step = $args->getStep();
+    $storage = new AutomationRunLogStorage();
+    $runLog = $storage->getAutomationRunLogByRunAndStepId($run->getId(), $step->getId());
+
+    if ($runLog === null) {
+      return [];
+    }
+    return $runLog->getData();
+  }
+
+  /**
+   * Schedule a progress run to sync the email sending status to the automation step.
+   * Normally, a progress run is executed immediately after sending; we're scheduling
+   * these runs to poll for the status if sync fails or email never sends (timeout),
+   * or if we need to wait for subscriber opt-in.
+   */
+  private function rerunLater(StepRunArgs $args, StepRunController $controller, NewsletterEntity $newsletter, SubscriberEntity $subscriber): void {
+    $runNumber = $args->getRunNumber();
+    $nextInterval = self::POLL_INTERVALS[$runNumber - 1] ?? 0;
+
+    // Use different intervals when retrying for opt-in.
+    if ($this->isOptInRequired($newsletter, $subscriber) === true) {
+      if ($runNumber > count(self::OPTIN_RETRY_INTERVALS)) {
+        $subscriberStatus = $subscriber->getStatus();
+        // translators: %s is the subscriber's status.
+        throw InvalidStateException::create()->withMessage(sprintf(__("Cannot send the email because the subscriber's status is '%s'.", 'mailpoet'), $subscriberStatus));
+      }
+      $nextInterval = self::OPTIN_RETRY_INTERVALS[$runNumber - 1];
+    }
+
     $controller->scheduleProgress(time() + $nextInterval);
   }
 
+  private function isOptInRequired(NewsletterEntity $newsletter, SubscriberEntity $subscriber): bool {
+    $subscriberStatus = $subscriber->getStatus();
+    return ($newsletter->getType() !== NewsletterEntity::TYPE_AUTOMATION_TRANSACTIONAL && $subscriberStatus !== SubscriberEntity::STATUS_SUBSCRIBED);
```

Comment:
Maybe just taste, but I find this hard to read. So much combined negativity.

i think that would make it easier for me:
```
if transactional return false;
return !subscribed
```

Reply:
No problem with less negativity. 😄

FTR I merely extracted existing code here. 😇